Implementation of NMS ONNX and TF V4.

ayermolo opened the pull request:
Summary:
This is somewhat of a messy, and probably controversial node.
It is needed for Object Detection post processing.
The core algorithm is basically ONNX runtime one. That being said there are some caveats. 
First one.
The TF models use NMSv4. Unlike ONNX spec definition it
1) Works only on 1 class.
2) Has second output of number of indices detected.

So this implementation tries to combine them.

Second caveat is that second output is being broadcasted internally within NMS node. This eliminates explicit broadcast node in the graph. 

Documentation:
https://github.com/onnx/onnx/blob/master/docs/Operators.md#NonMaxSuppression
https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/non-max-suppression-v4

[Optional Fixes #issue]

Test Plan:
External: Unit Tests
Internal: ONNX SSD, TF converted Mobilenet V2 SSD, DarkNet converted Yolo V2.

Related to https://github.com/pytorch/glow/issues/3944

ayermolo commented:
@jfix71 Is this something you guys would be interested?

jfix71 commented:
Hi @ayermolo -- yes this would be great to have. Sorry for the delay -- I will get to reviewing and landing this next week.

ayermolo commented:
@jfix71 Awesome, thanks!

ayermolo commented:
@jfix71 any updates? :)

jfix71 commented on tests/unittests/OperatorTest.cpp:
```diff
@@ -616,6 +616,411 @@ TEST_P(OperatorTest, where_element_wise_float) {
+  ENABLED_BACKENDS(Interpreter, CPU);
+  llvm::SmallVector<dim_t, 3> boxesDims = {1, 6, 4};
+  llvm::SmallVector<dim_t, 3> scoresDims = {1, 1, 6};
+  llvm::SmallVector<dim_t, 3> boxIndexesDIms = {1};
```
nit: size is 3, but initializing with 1 value.

jfix71 replied:
typo: `DIms`

jfix71 commented on tests/unittests/OperatorTest.cpp:
```diff
@@ -616,6 +616,411 @@ TEST_P(OperatorTest, where_element_wise_float) {
+  llvm::SmallVector<Box, 3> refResults = {
+      {0.5, 10.5, 1.0, 1.0}, {0.5, 0.5, 1.0, 1.0}, {0.5, 0.5, 1.0, 1.0}};
+  NMSMetaData metaData = {1, 3, 0.5, 0.4};
+  llvm::SmallVector<int32_t, 1> refNumSelected{2};
```
nit: for consistency with other uses, `refNumSelected = {2};`

jfix71 commented on tests/unittests/OperatorTest.cpp:
```diff
@@ -616,6 +616,411 @@ TEST_P(OperatorTest, where_element_wise_float) {
+      0.5, 10.5, 1.0, 1.0, 0.5, 10.6, 1.0, 1.0, 0.5, 100.5, 1.0, 1.0};
+
+  llvm::SmallVector<float, 6> classes = {0.9, 0.75, 0.6, 0.95, 0.5, 0.3};
+  llvm::SmallVector<int32_t, 3> boxIndices = {2};
```
nit: size is 3, but initializing with 1 value.

jfix71 commented on tests/unittests/OperatorTest.cpp:
```diff
@@ -616,6 +616,411 @@ TEST_P(OperatorTest, where_element_wise_float) {
+  auto *scores = createPlaceholderConditionallyQuantized(mod, DTy, scoresDims,
+                                                         "scores", false);
+
+  auto boxIndices = createPlaceholderConditionallyQuantized(
```
nit: `auto *boxIndices`

jfix71 commented on tests/unittests/OperatorTest.cpp:
```diff
@@ -616,6 +616,411 @@ TEST_P(OperatorTest, where_element_wise_float) {
+  }
+
+  auto *saveBoxes = F->createSave("saveBoxes", fltn2);
+  auto saveNumSelected =
```
nit: `auto *saveNumSelected`

jfix71 commented on tests/unittests/OperatorTest.cpp:
```diff
@@ -616,6 +616,411 @@ TEST_P(OperatorTest, where_element_wise_float) {
+  ENABLED_BACKENDS(Interpreter, CPU);
+  llvm::SmallVector<dim_t, 3> boxesDims = {6, 4};
+  llvm::SmallVector<dim_t, 3> scoresDims = {6};
+  llvm::SmallVector<dim_t, 3> boxIndexesDIms = {3};
```
typo: `DIms`

ayermolo commented:
Thanks for the comments, will integrate them in.

ayermolo commented:
@jfix71 ok, had time to circle back to it and updated.
Not sure why glang format fails. Passes on my system, and I didn't even touch the file it's complaining about:
modified:   tests/unittests/BackendTestUtils.h
The Pytorch error seems unrelated also:
E   ImportError: /var/lib/jenkins/workspace/torch_glow/torch_glow/_torch_glow.cpython-36m-x86_64-linux-gnu.so: undefined symbol: THPVariableClass

ayermolo commented:
@jfix71 Awesome, thanks!

jfix71 commented:
@ayermolo FYI for when this lands, I updated the OperatorTests here to use `CHECK_IF_ENABLED()` + blacklists instead of `ENABLED_BACKENDS()`, and the OnnxImporterTests to use `TEST_F(OnnxImporterTest, ...` instead of `TEST(onnx, ...`.

ayermolo commented:
@jfix71 Ah, ic. Thank you!
